### PR TITLE
Chat history

### PR DESCRIPTION
**Pull request by StasDash**
- Custom scrollbars
- Messages now starts from bottom
- Visual Changes
- Reaction was removed for rework
- Column limit is set to **140**

### CODE REVIEW

File: Client/Client.Qt/ChatWindow.cpp [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
+    mainLayout->setStretchFactor(1, 80);
 }
 
-ChatWindow::~ChatWindow()
-{
-    delete channelListWidget;
-    delete mainLayout;
-    delete chat;
-}
+void ChatWindow::resizeEvent(QResizeEvent*) { mainLayout->resize(width(), height()); }
```
Comment (fisyamaloy) [inline Client/Client.Qt/ChatWindow.cpp:24]: here need new line

File: Client/Client.Qt/Style/StyleBasic.cpp [hunk 2]
```diff
@@ -93,19 +100,23 @@ namespace
...
     px0  = ConvertScale(0, scale);
     px2  = ConvertScale(2, scale);
     px3  = ConvertScale(3, scale);
+    px4  = ConvertScale(4, scale);
     px5  = ConvertScale(5, scale);
     px8  = ConvertScale(8, scale);
     px10 = ConvertScale(10, scale);
+    px12 = ConvertScale(12, scale);
     px14 = ConvertScale(14, scale);
     px15 = ConvertScale(15, scale);
```
Comment (fisyamaloy) [inline Client/Client.Qt/Style/StyleBasic.cpp:121]: maybe, using some dynamic structure would be better then declaration of so many variables?

File: Client/Client.Qt/Style/StyleBasic.hpp [hunk 3]
```diff
@@ -55,6 +38,27 @@ struct FlatButton
...
+struct MessageWidget
+{
+    int                 radius;
+    Style::font         fontname;
+    Style::font         fontdate;
+    Style::font         fonttext;
+    int                 border;
+    Style::FlatButton   button;
+    Style::FlatTextEdit textedit;
+};
```
Comment (fisyamaloy) [inline Client/Client.Qt/Style/StyleBasic.hpp:60]: Don't forget about memory alignment

File: Client/Client.Qt/Widgets/ChatWidget.cpp [hunk 4]
```diff
@@ -1,61 +1,35 @@
...
-    auto* myItem =
-        new MessageWidget(std::move(messageText), std::move(userNameMessage), item, false);
-    myItem->setThisItem(item);
-    chatWidget->addItem(item);
-    chatWidget->setItemWidget(item, myItem);
+    chatWidget->addMessage(messageText, QDateTime::currentSecsSinceEpoch());
 }
 
-void ChatWidget::connectUi()
+void ChatWidget::newMessage(QString messageText, QString userNameMessage)
```
Comment (fisyamaloy) [inline Client/Client.Qt/Widgets/ChatWidget.cpp:27]: use reference &

File: Client/Client.Qt/Style/StyleBasic.cpp (hunk 2 above)
Comment (StasDash) [reply Client/Client.Qt/Style/StyleBasic.cpp:121]: No.
Only possible container for this is map, that may require from us to add additional check on valid data

File: Client/Client.Qt/Widgets/MessageWidget.cpp [hunk 5]
```diff
@@ -1,176 +1,85 @@
...
+#include "ChatHistory.hpp"
 #include "Style/Style.hpp"
 
-MessageWidget::MessageWidget(QString textMessage, QString nameOfUser, QListWidgetItem* Item,
-                             bool deletedMessage, QWidget* parent)
-    : QWidget(parent),
-      messageText(std::move(textMessage)),
-      userName(std::move(nameOfUser)),
-      dateTimeMessage(QDateTime::currentDateTime())
+MessageWidget::MessageWidget(QWidget* history, QString message, qint64 utc, QString user, const Style::MessageWidget& st)
```
Comment (fisyamaloy) [inline Client/Client.Qt/Widgets/MessageWidget.cpp:9]: if we use std::move, why in parameters not r-value objects?

File: Client/Client.Qt/Widgets/MessageWidget.cpp [hunk 6]
```diff
@@ -217,58 +125,48 @@ void MessageWidget::reactionChange(int index)
...
-    {
-        userName = QString::fromStdString(newUserName);
-        updateWidget();
-    }
+    dateTimeMessage.setSecsSinceEpoch(utc);
+    update();
 }
 
-void MessageWidget::setDateTime(QDateTime newDataTime)
+void MessageWidget::setReactionMap(std::map<int, int> newReactionMap)
```
Comment (fisyamaloy) [inline Client/Client.Qt/Widgets/MessageWidget.cpp:152]: here too

File: Client/Client.Qt/Widgets/MessageWidget.cpp [hunk 7]
```diff
@@ -217,58 +125,48 @@ void MessageWidget::reactionChange(int index)
...
-    reactionMap = std::move(newReactionMap);
-    updateWidget();
-}
+    if (!_fmtMessageText) return 0;
+    if (_messageFlags & MessageFlag::Deleted)
+    {
+        return _st.fontname->height + _st.radius * 2;
+    }
+    return _st.radius * 6 + _st.fontname->height + _fmtMessageText->document()->size().height();
+}
```
Comment (fisyamaloy) [inline Client/Client.Qt/Widgets/MessageWidget.cpp:172]: add new line

File: Client/Client.Qt/Widgets/MessageWidget.cpp (hunk 5 above)
Comment (StasDash) [reply Client/Client.Qt/Widgets/MessageWidget.cpp:9]: Clang-Tidy says its better

File: Client/Client.Qt/Widgets/MessageWidget.cpp (hunk 6 above)
Comment (StasDash) [reply Client/Client.Qt/Widgets/MessageWidget.cpp:153]: this thing is now on rework and not used

**Review by fisyamaloy — APPROVED** [review]

File: Client/Client.Qt/ChatWindow.cpp [hunk 8]
```diff
@@ -1,26 +1,24 @@
...
+    mainLayout        = std::make_unique<QSplitter>(this);
+    channelListWidget = std::make_unique<ChannelListWidget>();
+    chat              = std::make_unique<ChatWidget>();
+    mainLayout->setOrientation(Qt::Horizontal);
+    mainLayout->setChildrenCollapsible(false);
+    mainLayout->setContentsMargins(0, 0, 0, 0);
+    mainLayout->addWidget(channelListWidget.get());
+    mainLayout->addWidget(chat.get());
+    mainLayout->setStretchFactor(0, 20);
+    mainLayout->setStretchFactor(1, 80);
```
Comment (glaush) [inline Client/Client.Qt/ChatWindow.cpp:21]: We use labels for indicating class fields by code style.

File: Client/Client.Qt/ChatWindow.hpp [hunk 9]
```diff
@@ -7,10 +9,12 @@ class ChatWindow : public QWidget
...
+protected:
+    void resizeEvent(QResizeEvent* event) override;
 
 private:
-    QHBoxLayout* mainLayout;
-    ChannelListWidget* channelListWidget;
-    ChatWidget* chat;
+    std::unique_ptr<QSplitter>         mainLayout;
+    std::unique_ptr<ChannelListWidget> channelListWidget;
+    std::unique_ptr<ChatWidget>        chat;
```
Comment (glaush) [inline Client/Client.Qt/ChatWindow.hpp:19]: The same.

File: Client/Client.Qt/Style/Palette.hpp [hunk 10]
```diff
@@ -4,179 +4,204 @@
...
-namespace internal {
-
+namespace Style
+{
+namespace internal
+{
 void init_Palette(int scale);
 
-} // namespace internal
+}  // namespace internal
```
Comment (glaush) [inline Client/Client.Qt/Style/Palette.hpp:13]: For what purpose is it?

File: Client/Client.Qt/Style/Palette.hpp [hunk 11]
```diff
@@ -4,179 +4,204 @@
...
+void               apply(const palette& other);
+void               reset();
+int                indexOfColor(color c);
+
+struct row
+{
+    QLatin1String name;
+    QLatin1String value;
+    QLatin1String fallback;
+    QLatin1String description;
```
Comment (glaush) [inline Client/Client.Qt/Style/Palette.hpp:182]: The same. (labels for struct fields)

File: Client/Client.Qt/Widgets/ChannelListWidget.cpp [hunk 12]
```diff
@@ -3,14 +3,14 @@
...
     addChannelButton   = std::make_unique<FlatButton>(this, "+");
 
     setContentsMargins(0, 0, 0, 0);
     addChannelButton->setMinimumWidth(Style::valueDPIScale(200));
     vBoxLayout->addWidget(addChannelButton.get());
-    vBoxLayout->addWidget(channelList);
+    vBoxLayout->addWidget(channelList.get());
     channelList->setFrameShape(QFrame::NoFrame);
 
     addChannelButton->setClickCallback([this]() { addChannels(); });
```
Comment (glaush) [inline Client/Client.Qt/Widgets/ChannelListWidget.cpp:16]: The same. (labels for struct fields)

File: Client/Client.Qt/Widgets/ChannelListWidget.hpp [hunk 13]
```diff
@@ -15,14 +15,13 @@ class ChannelListWidget : public QWidget
...
 
 private:
-    std::unique_ptr<FlatButton> addChannelButton;
-    std::unique_ptr<QVBoxLayout> vBoxLayout;
-    ChannelListWindow* listWidgetChannels;
-    ListWidget* channelList;
+    std::unique_ptr<FlatButton>        addChannelButton;
+    std::unique_ptr<ChannelListWindow> listWidgetChannels;
+    std::unique_ptr<QVBoxLayout>       vBoxLayout;
+    std::shared_ptr<ListWidget>        channelList;
```
Comment (glaush) [inline Client/Client.Qt/Widgets/ChannelListWidget.hpp:26]: The same. (labels for struct fields)

File: Client/Client.Qt/Widgets/ChannelListWindow.hpp [hunk 14]
```diff
@@ -34,9 +35,9 @@ public slots:
...
-    std::unique_ptr<ListWidget> channelList;
-    std::unique_ptr<FlatButton> addChannelButton;
-    std::unique_ptr<FlatButton> updateChannelButton;
-    std::unique_ptr<QVBoxLayout> vBoxLayout;
-    ListWidget* channelListMainWindow;
+    std::unique_ptr<ListWidget>            channelList;
+    std::unique_ptr<FlatButton>            addChannelButton;
+    std::unique_ptr<FlatButton>            updateChannelButton;
+    std::unique_ptr<QVBoxLayout>           vBoxLayout;
+    std::shared_ptr<ListWidget>            channelListMainWindow;
```
Comment (glaush) [inline Client/Client.Qt/Widgets/ChannelListWindow.hpp:42]: The same. (labels for struct fields)

File: Client/Client.Qt/Widgets/ChatHistory.cpp [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+        _scrollArea->scrollToY(_messageList.back()->pos().y());
+    }
+}
+
+void ChatHistory::resizeVisible()
+{
+    if (_alreadyScrolling) return;
+    if (_messageList.empty())
+    {
+        _left = -1;
```
Comment (glaush) [inline Client/Client.Qt/Widgets/ChatHistory.cpp:109]: We need named constants for this case.

File: Client/Client.Qt/Widgets/ChatHistory.cpp [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        return;
+    }
+    auto [left, right] = findVisible();
+    int  width         = this->width() - 25;
```
Comment (glaush) [inline Client/Client.Qt/Widgets/ChatHistory.cpp:113]: We need named constants for this case.

File: Client/Client.Qt/Widgets/InputFields.cpp [hunk 17]
```diff
@@ -103,52 +102,44 @@ FlatInput::FlatInput(const QString& placeholder, bool password, QWidget* parent)
...
     painter.setPen(Qt::NoPen);
-    painter.drawRect(QRectF(0, 0, width(), height()).marginsRemoved(QMarginsF(2, 2, 2, 2)));
+
+    painter.setBrush(_st.bg);
+    painter.drawRect(rect());
 
     QTextEdit::paintEvent(event);
 }
+
+void FlatTextEdit::resizeEvent(QResizeEvent* resizeEvent) { QTextEdit::resizeEvent(resizeEvent); }
```
Comment (glaush) [inline Client/Client.Qt/Widgets/InputFields.cpp:145]: Enter at the EOF.

File: Client/Client.Qt/Widgets/MessageWidget.cpp [hunk 18]
```diff
@@ -1,274 +1,129 @@
 ﻿#include "MessageWidget.hpp"
 
+#include <QPainter>
+#include <QtEvents>
+#include <utility>
+
+#include "ChatHistory.hpp"
 #include "Style/Style.hpp"
```
Comment (glaush) [inline Client/Client.Qt/Widgets/MessageWidget.cpp:8]: Place the headers in some order, but not in a jumble.

**Review by glaush — DISMISSED** [review]

File: Client/Client.Qt/Style/Palette.hpp (hunk 10 above)
Comment (StasDash) [reply Client/Client.Qt/Style/Palette.hpp:13]: Clang format

File: Client/Client.Qt/Widgets/ChatHistory.cpp (hunk 15 above)
Comment (StasDash) [reply Client/Client.Qt/Widgets/ChatHistory.cpp:109]: There is no constants

File: Client/Client.Qt/Widgets/MessageWidget.cpp (hunk 18 above)
Comment (StasDash) [reply Client/Client.Qt/Widgets/MessageWidget.cpp:8]: Clang Format

File: Client/Client.Qt/Style/Palette.hpp (hunk 11 above)
Comment (StasDash) [reply Client/Client.Qt/Style/Palette.hpp:182]: Aren't labels for private data?

File: DataAccess/DataAccess.Static/UsersAmountFinder.hpp [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    virtual ~UsersAmountFinder() = default;
+
+    std::uint16_t findUsersAmountWithSameTableAttribute(const std::string& condition) const;
+
+    std::uint16_t findUsersAmountWithSameLogin(const std::string& login) const;
+
+    std::uint16_t findUsersAmountWithSameEmail(const std::string& email) const;
+
+    std::uint16_t findUsersAmountWithAllSameData(const Network::RegistrationInfo& ri) const;
+};
```
Comment (Svidunovich) [inline DataAccess/DataAccess.Static/UsersAmountFinder.hpp:26]: Add documentation for this file plz

Comment (StasDash) [reply DataAccess/DataAccess.Static/UsersAmountFinder.hpp:26]: This is not my file

File: Client/Client.Qt/Widgets/ChannelListWindow.hpp [hunk 20]
```diff
@@ -33,10 +33,10 @@ public slots:
...
-    std::unique_ptr<FlatButton> addChannelButton;
-    std::unique_ptr<FlatButton> updateChannelButton;
-    std::unique_ptr<QVBoxLayout> vBoxLayout;
-    ListWidget* channelListMainWindow;
+    inline static std::vector<std::string> _channelNames{};
+    std::unique_ptr<ListWidget>            _channelList;
+    std::unique_ptr<FlatButton>            _addChannelButton;
+    std::unique_ptr<FlatButton>            _updateChannelButton;
+    std::unique_ptr<QVBoxLayout>           _vBoxLayout;
+    std::shared_ptr<ListWidget>            _channelListMainWindow;
```
Comment (GuskoMaksim) [inline Client/Client.Qt/Widgets/ChannelListWindow.hpp:41]: In this case, camelcase is used, without the underscore

File: Client/Client.Qt/Widgets/MessageWidget.hpp [hunk 21]
```diff
@@ -116,50 +97,16 @@ private slots:
...
-    void clearMessage();
-};
+    std::unique_ptr<FlatTextEdit> _fmtMessageText;
+    std::unique_ptr<FlatButton>   _deleteBtn;
+    std::int32_t                  _index = 0;
+    QString                       _messageText;
+    QString                       _username;
+    QDateTime                     _datetime;
+    MessageFlags                  _messageFlags;
+    const Style::MessageWidget&   _st;
```
Comment (GuskoMaksim) [inline Client/Client.Qt/Widgets/MessageWidget.hpp:109]: and here too

**Review by GuskoMaksim — CHANGES_REQUESTED** [review]

**Review by Svidunovich — APPROVED** [review]

**Pull request merged**